test/toolbox_image.sh

```diff
@@ -12,243 +12,65 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
-        exit 1
-    fi
-
-    overlayfs_dir="$(mktemp -d)"
-    rootfs_dir="$(mktemp -d)"
-
-    mount -t tmpfs none "${overlayfs_dir}"
-    mkdir "${overlayfs_dir}/rom"
-    mkdir "${overlayfs_dir}/up"
-    mkdir "${overlayfs_dir}/work"
-
-    mount "${ROOTFS_IMG}" "${overlayfs_dir}/rom"
-    mount -t overlay overlay \
-          -o "lowerdir=${overlayfs_dir}/rom,upperdir=${overlayfs_dir}/up,workdir=${overlayfs_dir}/work" \
-          "${rootfs_dir}"
-
-    mkdir -p "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}"
-    tar xvfJ "${TEST_UPDATE_ROOTFS_FILE}" -C "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}" \
-        > /dev/null 2> /dev/null
-
-    update_mount="$(mktemp -d)"
-
-    touch "${rootfs_dir}/${ARM_EMU_BIN}"
-    mount --bind -o ro "${ARM_EMU_BIN}" "${rootfs_dir}/${ARM_EMU_BIN}"
-
-    mount --bind /proc "${rootfs_dir}/proc" 1> /dev/null
-    ln -s ../proc/self/mounts "${rootfs_dir}/etc/mtab"
-
-    if ! grep -qE "/dev.*devtmpfs" /proc/mounts; then
-        mount -t devtmpfs none "/dev"
-        is_dev_setup_mounted=true
-    fi
-
-    mount -t devtmpfs none "${rootfs_dir}/dev"
-
-    dd if=/dev/zero of="${rootfs_dir}/${TMP_TEST_IMAGE_FILE}" bs=1 count=0 seek=128M 2> /dev/null
-
-    create_dummy_storage_device
+    set -e
+    toolbox_root_dir="$(mktemp -d)"
+    setup_armhf_emulation "${toolbox_image}" "${toolbox_root_dir}"
 
-    sfdisk -d "${LOOP_STORAGE_DEVICE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}"
-
-    # Add the partition labels to the partition file, they will be used as label
-    sed -i "s|${LOOP_STORAGE_DEVICE}p1.*|&, name=boot|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
-    sed -i "s|${LOOP_STORAGE_DEVICE}p2.*|&, name=root|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
-    sed -i "s|${LOOP_STORAGE_DEVICE}p3.*|&, name=user|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
+    dd if=/dev/zero of="${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}" bs=1 count=0 seek=128M
 }
 
 teardown()
 {
-    mounts="${rootfs_dir} ${overlayfs_dir}/rom ${overlayfs_dir}"
-
-    if [ ! -d "${rootfs_dir}" ]; then
-        return
-    fi
-
-    for partition_table_file in "${rootfs_dir}${PARTITION_TABLE_FILE}"*; do
-        unlink "${partition_table_file}"
-    done
-
-    if [ -d "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}" ]; then
-        rm -rf "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}"
+    set -e
+    if [ -f "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}" ]; then
+        unlink "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}"
     fi
 
-    if [ -d "${update_mount}" ]; then
-        rm -r "${update_mount:?}"
-    fi
+    teardown_armhf_emulation "${toolbox_root_dir}"
 
-    if [ -b "${LOOP_STORAGE_DEVICE}" ]; then
-        losetup -d "${LOOP_STORAGE_DEVICE}"
+    if grep -qE "${toolbox_root_dir}" "/proc/mounts"; then
+        umount "${toolbox_root_dir}"
     fi
 
-    if [ -f "${rootfs_dir}${STORAGE_DEVICE_IMG}" ]; then
-        unlink "${rootfs_dir}${STORAGE_DEVICE_IMG}"
+    if [ -d "${toolbox_root_dir}" ]; then
+        rm -rf "${toolbox_root_dir:?}"
```

Comment:
i'm slightly uneasy with this command, as the ? does protect you from empty variable (rm -rf NULL exits also i believe) it does not protect from rm -rf "/".

we could say, we add the -t="um-toolbox.XXXXXX" parameter to mktemp and add the test `&& [ "${toolbox_root_dir#"um-toolbox.."}" != "um-toolbox." ]`

this ensures that we only ever delete the directory if it at least contains our directory template.

While i'd leave the -d check in place, that would be then optional as well, as we create something, we delete it.